File: attendance/adapter/js/presence-rules.js
'use strict';

/**
 * Presence Rules
 *
 * ISOLATED MODULE - Business rules for determining attendance status.
 *
 * Current rule:
 *   The meeting is split in two halves (midpoint).
 *   If a participant was present >= 80% of a half, they count as "present" for that half.
 *
 * !! MODIFY THIS FILE when attendance rules change !!
 * !! Nothing else in the codebase depends on HOW the decision is made, only on the result !!
 */
const PresenceRules = {

    /** Minimum fraction of a half to be considered "present" for that half */
    threshold: 0.8,

    /**
     * Determine attendance status for a participant.
     *
     * @param {number} minutesFirstHalf  - Minutes present in first half
     * @param {number} minutesSecondHalf - Minutes present in second half
     * @param {number} durationFirstHalf - Total duration of first half (minutes)
     * @param {number} durationSecondHalf - Total duration of second half (minutes)
     * @returns {string} "presente" | "prima_meta" | "seconda_meta" | "assente"
     */
    determine(minutesFirstHalf, minutesSecondHalf, durationFirstHalf, durationSecondHalf, threshold) {
        const t = (threshold !== undefined) ? threshold : this.threshold;

        const firstOk = durationFirstHalf > 0 &&
            (minutesFirstHalf / durationFirstHalf) >= t;

        const secondOk = durationSecondHalf > 0 &&
            (minutesSecondHalf / durationSecondHalf) >= t;

        if (firstOk && secondOk) return 'presente';
        if (firstOk) return 'prima_meta';
        if (secondOk) return 'seconda_meta';
        return 'assente';
    }
};

File: backend/attendance_normalization/presence_rules.py
"""Presence status rules migrated from ``attendance/adapter/js/presence-rules.js``.

This module intentionally mirrors the current browser logic without adding
extra heuristics. The goal of this first migration step is to make the
decision function explicit, testable, and reusable from Python.
"""

STATUS_PRESENT = "presente"
STATUS_FIRST_HALF = "prima_meta"
STATUS_SECOND_HALF = "seconda_meta"
STATUS_ABSENT = "assente"

DEFAULT_THRESHOLD = 0.8


def determine_presence_status(
    minutes_first_half: float,
    minutes_second_half: float,
    duration_first_half: float,
    duration_second_half: float,
    threshold: float = DEFAULT_THRESHOLD,
) -> str:
    """Classify presence exactly like the current JavaScript rule set.

    A participant counts for a half only if:
    - that half has a positive duration;
    - present minutes / half duration >= threshold.

    The final status is:
    - ``presente`` when both halves pass the threshold;
    - ``prima_meta`` when only the first half passes;
    - ``seconda_meta`` when only the second half passes;
    - ``assente`` otherwise.
    """

    first_ok = duration_first_half > 0 and (
        minutes_first_half / duration_first_half
    ) >= threshold
    second_ok = duration_second_half > 0 and (
        minutes_second_half / duration_second_half
    ) >= threshold

    if first_ok and second_ok:
        return STATUS_PRESENT
    if first_ok:
        return STATUS_FIRST_HALF
    if second_ok:
        return STATUS_SECOND_HALF
    return STATUS_ABSENT
